Python/compile.c

```diff
@@ -8250,23 +8245,19 @@ optimize_basic_block(struct compiler *c, basicblock *bb, PyObject *consts)
             case JUMP_IF_FALSE_OR_POP:
                 switch(target->i_opcode) {
                     case POP_JUMP_IF_FALSE:
-                        if (inst->i_lineno == target->i_lineno) {
-                            *inst = *target;
-                            i--;
-                        }
+                        i -= jump_thread(inst, target, POP_JUMP_IF_FALSE);
                         break;
                     case JUMP_ABSOLUTE:
                     case JUMP_FORWARD:
                     case JUMP_IF_FALSE_OR_POP:
-                        if (inst->i_lineno == target->i_lineno &&
-                            inst->i_target != target->i_target) {
-                            inst->i_target = target->i_target;
-                            i--;
-                        }
+                        i -= jump_thread(inst, target, JUMP_IF_FALSE_OR_POP);
                         break;
                     case JUMP_IF_TRUE_OR_POP:
-                        assert (inst->i_target->b_iused == 1);
+                    case POP_JUMP_IF_TRUE:
                         if (inst->i_lineno == target->i_lineno) {
+                            // We don't need to bother checking for loops here,
+                            // since inst->i_target can't possibly be equal to
+                            // inst->i_target->b_next:
```

Comment:
Why can't it be equal?
Could you use `jump_thread` here as well? The redundant check is harmless.

Reply:
We don’t want to use `jump_thread` here. Unlike all the others, these cases don’t actually take the second jump; they “fall through” to the block following their target (notice that that `inst->i_target` becomes `inst-i_target->b_next`, not `target->i_target`). 

As to the need to check for loops: I don’t see how a block could possibly fall through to itself. Should I add an assert, just to be sure?

Reply:
`b_next`  shouldn't point to itself (I think if it did we would have an infinite loop somewhere).
Maybe change the comment from
"inst->i_target can't possibly be equal to inst->i_target->b_next" to
"a block's b_next cannot point to itself".
An assertion would do no harm.